### PR TITLE
Added `ARTrustRegion`

### PR DESCRIPTION
**Pull request by Jay-sanjay**
This PR is related to  - https://github.com/JuliaSmoothOptimizers/AdaptiveRegularization.jl/issues/92

closes - https://github.com/JuliaSmoothOptimizers/AdaptiveRegularization.jl/issues/92

### CODE REVIEW

File: src/trust-region/ar-trust-region.jl [hunk 1]
```diff
@@ -0,0 +1,110 @@
+export ARTrustRegion
+
+# Already exists in SolverTools.jl
+"Exception type raised in case of error."
+mutable struct TrustRegionException <: Exception
+  msg::String
+end
```
Comment (tmigot) [inline src/trust-region/ar-trust-region.jl:7]:
```suggestion
```

Comment (tmigot) [reply src/trust-region/ar-trust-region.jl:7]: As the comment says this already exists.

File: src/trust-region/ar-trust-region.jl [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+- `acceptance_threshold::T`: Ratio over which the step is successful. Default `T(0.1)`.
+- `increase_threshold::T`: Ratio over which we increase `α`. Default `T(0.75)`.
+- `reduce_threshold::T`: Ratio under which we decrease `α`. Default `T(0.1)`.
+- `increase_factor::T`: Factor of increase of `α`. Default `T(5.0)`.
+- `decrease_factor::T`: Factor of decrease of `α`. Default `T(0.1)`.
+- `max_unsuccinarow::Int`: Limit on the number of successive unsucessful iterations. Default `30`.
+
+Returns a `ARTrustRegion` structure.
+
+This can be compared to https://github.com/JuliaSmoothOptimizers/SolverTools.jl/blob/main/src/trust-region/basic-trust-region.jl
```
Comment (tmigot) [inline src/trust-region/ar-trust-region.jl:24]:
```suggestion
This can be compared to `TrustRegion` or `TRONTrustRegion`.
```

**Review by tmigot — CHANGES_REQUESTED** [review]
Could you also add a unit test for the functions defined here?

**Comment by Jay-sanjay** [comment]
HI @tmigot . Sir I was a bit confused with how the tests must look like. Should they just check what all things it is returning like the below once or something else is required.

```
  @testset "ARTrustRegion" begin
    Δ₀ = 0.8
    ar = ARTrustRegion(0.8)
    @test ar.α₀ == 0.8 
    @test ar.α == 0.8
    @test ar.max_α ≈ 1 / sqrt(eps(typeof(0.8)))
    @test ar.acceptance_threshold == 0.1
    @test ar.increase_threshold == 0.75
    @test ar.reduce_threshold == 0.1
    @test ar.increase_factor == 5.0
    @test ar.decrease_factor == 0.1
    @test ar.large_decrease_factor == 0.01
    @test ar.max_unsuccinarow == 30
  end
  ```
  
And also to just let you know, I have also asked you the same thing on slack, there we can take on a conversation for this :)
thanks !

**Comment by tmigot** [comment]
I replied on Slack. Unit tests should check the constructors (as you suggested) but also add check for the added function.

**Review by tmigot — APPROVED** [review]
Thanks for the PR @Jay-sanjay ! I took the liberty to add some more tests and it looks good.

**Comment by Jay-sanjay** [comment]
Thanks for the review @tmigot , also I was a bit curious that how come codecov passed all checks by adding extra test-cases (like 2 more extra testcases). Did they covered any edge cases ? Or what else ?

**Comment by tmigot** [comment]
Yes, it was covering edge cases in the compute_r function
